Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift

```diff
@@ -1257,100 +1333,128 @@ private final class PubGrubPackageContainer {
+                    self.getContainer(for: identifier, completion: completion)
+                }
+            }
+        } else {
+            // Otherwise, fetch the container from the provider
+            self.provider.getContainer(for: identifier, skipUpdate: skipUpdate, on: self.queue) { result in
+                switch result {
+                case .failure(let error):
+                    return completion(.failure(error))
+                case .success(let container):
+                    // only cache positive results
+                    let pubGrubContainer = PubGrubPackageContainer(underlying: container, pinsMap: self.pinsMap, queue: self.queue)
+                    self.containersCache[identifier] = pubGrubContainer
+                    return completion(.success(pubGrubContainer))
+                }
             }
-
-            // Otherwise, fetch the container synchronously.
-            let container = try tsc_await { provider.getContainer(for: identifier, skipUpdate: skipUpdate, completion: $0) }
-            let pubGrubContainer = PubGrubPackageContainer(container, pinsMap: pinsMap)
-            self._fetchedContainers[identifier] = .success(pubGrubContainer)
-            return pubGrubContainer
         }
     }
 
     /// Starts prefetching the given containers.
     func prefetch(containers identifiers: [PackageReference]) {
-        fetchCondition.whileLocked {
-            // Process each container.
-            for identifier in identifiers {
-                // Skip if we're already have this container or are pre-fetching it.
-                guard _fetchedContainers[identifier] == nil,
-                    !_prefetchingContainers.contains(identifier) else {
-                        continue
-                }
-
-                // Otherwise, record that we're prefetching this container.
-                _prefetchingContainers.insert(identifier)
-
-                provider.getContainer(for: identifier, skipUpdate: skipUpdate) { container in
-                    DispatchQueue.global().async {
-                        self.fetchCondition.whileLocked {
-                            // Update the structures and signal any thread waiting
-                            // on prefetching to finish.
-                            let pubGrubContainer = container.map {
-                                PubGrubPackageContainer($0, pinsMap: self.pinsMap)
-                            }
-                            self._fetchedContainers[identifier] = pubGrubContainer
-                            self._prefetchingContainers.remove(identifier)
-                            self.fetchCondition.signal()
-                        }
-                    }
+        // Process each container.
+        for identifier in identifiers {
+            if self.prefetches[identifier] != nil {
+                continue
+            }
+            self.prefetches[identifier] = DispatchGroup()
+            self.prefetches[identifier]?.enter()
```

Comment:
I’m on the fence about whether optional unwrap is better here rather than force unwrap.  I dislike force unwrap, but in this case, I think any logic error involving the prefect dispatch groups would result in a hang, which isn’t any better.  It’s a small point, just an odd pattern.

Reply:
in this specific case, there is no risk of hang, with the code that waits on the group looks like:

```
      if let prefetchSync = self.prefetches[identifier] {
            // If this container is already being prefetched, wait for that to complete
            self.queue.async {
                prefetchSync.wait()

                if let container = self.containersCache[identifier] {
                    // should be in the cache once prefetch completed
                    return completion(.success(container))
                } else {
                    // if prefetch failed, remove from list of prefetches and try again
                    self.prefetches[identifier] = nil
                    self.getContainer(for: identifier, completion: completion)
                }
            }
        }
```

so if "enter" is not called, the "wait" will crash